eddy/src/ciphertext.rs

```diff
@@ -0,0 +1,52 @@
+use crate::{decryption_share::Verified, limb, DecryptionShare, DecryptionTable, Value};
+
+/// A flow encryption ciphertext.
+pub struct Ciphertext {
+    pub(crate) c0: limb::Ciphertext,
+    pub(crate) c1: limb::Ciphertext,
+    pub(crate) c2: limb::Ciphertext,
+    pub(crate) c3: limb::Ciphertext,
+}
+
+impl Ciphertext {
+    /// Assumes decryption shares are verified already.
+    pub async fn decrypt(
+        &self,
+        shares: Vec<DecryptionShare<Verified>>,
+        table: &dyn DecryptionTable,
+    ) -> anyhow::Result<Value> {
+        let limb0_shares = shares.iter().map(|s| &s.share0).collect();
+        let limb1_shares = shares.iter().map(|s| &s.share1).collect();
+        let limb2_shares = shares.iter().map(|s| &s.share2).collect();
+        let limb3_shares = shares.iter().map(|s| &s.share3).collect();
+
+        let decryption0 = self.c0.decrypt(limb0_shares);
+        let decryption1 = self.c1.decrypt(limb1_shares);
+        let decryption2 = self.c2.decrypt(limb2_shares);
+        let decryption3 = self.c3.decrypt(limb3_shares);
+
+        let value0 = table
+            .lookup(decryption0.compress().0)
+            .await?
+            .ok_or_else(|| anyhow::anyhow!("could not find value in LUT"))?;
+        let value1 = table
+            .lookup(decryption1.compress().0)
+            .await?
+            .ok_or_else(|| anyhow::anyhow!("could not find value in LUT"))?;
+        let value2 = table
+            .lookup(decryption2.compress().0)
+            .await?
+            .ok_or_else(|| anyhow::anyhow!("could not find value in LUT"))?;
+        let value3 = table
+            .lookup(decryption3.compress().0)
+            .await?
+            .ok_or_else(|| anyhow::anyhow!("could not find value in LUT"))?;
```

Comment:
Mathematically, decryption can only fail if the value is not present in the lookup table, and while the point of the value integrity proofs is to guarantee that this cannot occur, someone could add too many ciphertexts at once, or similar, so it's important to model.

But in practice, the lookup table may not be entirely resident in memory, and might involve access to I/O, which is (1) async and (2) fallible.  So there are two kinds of errors: errors from the underlying storage, and a failed lookup.

The lookup trait returns a `Result<Option<u32>, anyhow::Error>`, with an outer `Result` allowing indication of an arbitrary I/O error, and the inner `Option` signaling whether or not the value was present.  This is a good API for external users to plug into, since they can put whatever I/O errors they want in the `Err` variant, and it's easy to signal missing values with `None`.

Here in decryption, though, we probably want a way to allow users to distinguish whether the error was due to underlying I/O issues, or whether a value went out of range.  If we use `anyhow!` to make a stringy error, we can't do this.

There are two options:

1. We could change the return type of decryption to `Result<Option<Value>, anyhow::Error>`, with the inner `Option` signaling whether the value was too large or not;
2. We could leave the return type as `Result<Value, anyhow::Error>`, but create a `LimbRangeError` struct, and use `thiserror` to derive an `Error` implementation for it, then do `.ok_or(LimbRangeError)?` here.

If we do the first option, all downstream users have to manually transform the inner `Option`. If we do the second option, downstream users that don't care can bubble up a generic error, and downstream users that do care can use `downcast_ref::<LimbRangeError>()` to handle it specially. 

So option (2) seems better to me.